### ISSUE TITLE
should FetchEvent.request ever be a range request if we cannot verify if the underlying resource is the same?

### ISSUE THREAD

**Issue by wanderview** [w3c/ServiceWorker#1201]
It seems that in firefox we avoid using partial response and range requests if:

1. The server cannot support range requests
2. The underlying resource has changed since the previous part of the resource was received

See:

https://searchfox.org/mozilla-central/source/netwerk/base/nsIResumableChannel.idl#13

If we allow range requests through FetchEvent.request, how can we ever know if the underlying resource has changed?  Normally we depend on the http cache to detect if a change has taken place, but we don't have that here.  The service worker could return completely different data each time.

Does chrome have similar "detect if something changed" behavior for range requests?

**Comment by wanderview** [comment]
As far as I can tell the fetch spec does not indicate when the Range header should be added to the request.  In gecko we only add it in "network or cache" path here:

https://searchfox.org/mozilla-central/source/netwerk/protocol/http/nsHttpChannel.cpp#4208

@annevk, is this something fetch spec should include?

**Comment by wanderview** [comment]
Gecko bug for this issue:

https://bugzilla.mozilla.org/show_bug.cgi?id=1404041

**Comment by annevk** [comment]
@wanderview yeah, @jakearchibald has some work-in-progress on range requests here: https://github.com/whatwg/fetch/pull/560.

**Comment by jakearchibald** [comment]
Yeah, I need to pick that back up again

**Comment by wanderview** [comment]
That PR looks useful to clarify some things, but does it address the checking for change question?  Maybe firefox is more conservative here, but we never use range request unless we can have some assurance that we're continuing the same resource from before.  All the logic that enables that is in the http cache which sits below the service worker.

**Comment by jakearchibald** [comment]
@wanderview here's the other part of it https://github.com/whatwg/html/pull/2814.

I'll also work on how to resume downloads, which we also need for background fetch.

You might be interested in my tests https://github.com/jakearchibald/range-request-test - Firefox is the only browser that does well.

**Comment by wanderview** [comment]
Thanks, I was also looking at:

https://docs.google.com/document/d/1SphP-WNxqzZrSv_6ApC9_FpM-m_tLzm57oL3SNGg-40/edit#

**Comment by monica-ch** [comment]
Bug triage: Service Worker handles the range requests but we might need to prevent range headers. See https://github.com/whatwg/fetch/issues/1906 for more information
